routing/http/types/ndjson/provider.go

```diff
@@ -0,0 +1,96 @@
+package ndjson
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/ipfs/go-libipfs/routing/http/types"
+	"github.com/ipfs/go-libipfs/routing/http/types/iter"
+)
+
+type readProvidersResponseIter struct {
+	iter.Iter[types.UnknownProviderRecord]
+}
+
+func NewReadProvidersResponseIter(r io.Reader) *readProvidersResponseIter {
+	return &readProvidersResponseIter{Iter: iter.FromReaderJSON[types.UnknownProviderRecord](r)}
+}
+
+func (p *readProvidersResponseIter) Next() (types.ProviderResponse, bool, error) {
+	v, ok, err := p.Iter.Next()
+	if err != nil {
+		return nil, false, err
+	}
+	if !ok {
+		return nil, false, nil
+	}
```

Comment:
Iterator chaining is a good example to see how this can be simplified using only two return params:

```go
v, err := p.iter.Next()
if err != nil {
   return nil, err
}
...
```

Reply:
btw this is an example of why you should use `errors.Is(err, io.EOF)`, I almost always wrap these kinds of errors with some context which would break `err == io.EOF` checks

Reply:
`io.EOF` is intended to be used like `err == io.EOF` and must not be wrapped because it is a sentinel error: https://github.com/golang/go/issues/39155

Reply:
even worse then that you cant use standard idioms with it

Reply:
> btw this is an example of why you should use `errors.Is(err, io.EOF)`, I almost always wrap these kinds of errors with some context which would break `err == io.EOF` checks

It is widely accepted that `io.EOF` must not be wrapped:
> EOF is the error returned by Read when no more input is available. (Read must return EOF itself, not an error wrapping EOF, because callers will test for EOF using ==.) Functions should return EOF only to signal a graceful end of input. If the EOF occurs unexpectedly in a structured data stream, the appropriate error is either ErrUnexpectedEOF or some other error giving more detail.

Reply:
> EOF is the error returned by Read when no more input is available

This is not Read()...